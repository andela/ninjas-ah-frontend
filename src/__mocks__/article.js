export default (Object.article = {
  title: 'my title well done',
  description: 'description of the article: well',
  body: 'body of the article: well'
});
