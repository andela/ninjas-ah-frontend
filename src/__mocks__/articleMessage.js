export default (Object.message = { message: 'article message' });
