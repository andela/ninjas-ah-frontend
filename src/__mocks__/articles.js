export default (Object.articles = [
  {
    author: { username: 'nivelo', bio: null, image: null },
    body: 'body',
    coverUrl: 'v1562366643/ninjas/ezzmjn0ak9b5gyzuzo6j.png',
    createdAt: '2019-07-05T22:43:57.384Z',
    description: 'ijihjhj',
    dislikes: null,
    favorited: false,
    favoritesCount: 0,
    id: 3,
    likes: null,
    rating: 0,
    readTime: '0',
    slug: 'how-to-conditionally-build-an-object-in-javascript-with-es6-m5mjxqor8sh',
    status: 'published',
    tagList: null,
    title: 'How to conditionally build an object in JavaScript with ES6',
    updatedAt: '2019-07-09T12:36:34.033Z',
    userId: 1
  }
]);
