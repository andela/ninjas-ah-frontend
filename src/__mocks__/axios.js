const mockAxios = jest.genMockFromModule('axios');

export const resolvedRequest = {
  data: {},
  status: 200,
  statusText: 'OK',
  headers: {},
  config: {}
};

export const rejectedRequest = {
  response: {
    data: { errors: { message: 'errors' } },
    status: 500,
    statusText: 'ERROR',
    headers: {},
    config: {}
  }
};

export default {
  get: jest.fn(() => Promise.resolve({ ...resolvedRequest })),
  post: jest.fn(() => Promise.resolve({ ...resolvedRequest, status: 201 })),
  put: jest.fn(() => Promise.resolve({ ...resolvedRequest })),
  delete: jest.fn(() => Promise.resolve({ ...resolvedRequest })),
  patch: jest.fn(() => Promise.resolve({ ...resolvedRequest })),
  create: jest.fn(() => mockAxios)
};
