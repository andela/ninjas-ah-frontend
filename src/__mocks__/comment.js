export default (Object.comment = { body: 'my title well done' });
