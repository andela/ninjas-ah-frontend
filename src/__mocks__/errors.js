export default (Object.errors = { message: 'can not updated article' });
