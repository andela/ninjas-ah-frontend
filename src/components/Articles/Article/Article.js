import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import 'dotenv/config';
import MetaTags from 'react-meta-tags';
import LazyLoad from 'react-lazyload';
import { Editor, EditorState, convertFromRaw } from 'draft-js';
import { fetchOneArticle } from '../../../actions';
import avatar from '../../../assets/images/user.png';
import timeStamp from '../../../helpers/timeStamp';
import { NotFound } from '../../common';
import Layout from '../../Layout';
import './Article.scss';

const { REACT_APP_IMAGE_BASE_URL } = process.env;

export class Article extends Component {
  state = {
    article: {},
    imageRectangle:
      'c_fill,g_auto,h_350,w_970/b_rgb:000000,e_gradient_fade,y_-0.20/c_scale,co_rgb:ffffff',
    editorState: EditorState.createEmpty()
  };

  async getSpecificArticle(slug) {
    const { fetchOneArticle } = this.props;
    await fetchOneArticle(slug);

    const { article } = this.props;
    const editorState = article.body
      ? EditorState.createWithContent(convertFromRaw(JSON.parse(article.body)))
      : EditorState.createEmpty();

    return this.setState({ article, editorState });
  }

  componentDidMount() {
    const { match: { params: { slug } } } = this.props;
    this.getSpecificArticle(slug);
  }

  render() {
    const { imageRectangle, article, editorState } = this.state;
    return (
      <Layout>
        <div id="article">
          {article && article.id ? (
            <div className="row">
              <MetaTags>
                <title> {article.title || 'Welcome'} - Authors Haven</title>
                <meta
                  name={article.description || 'Authors Haven'}
                  content={
                    article.content
                      ? article.content.replace(/<[^>]*>?/gm, '').trim(60)
                      : 'Authors Haven'
                  }
                />
                <meta property="og:title" content={article.title || 'Authors Haven'} />
                <meta property="og:image" content={article.coverUrl || ''} />
              </MetaTags>
              <LazyLoad height={350}>
                {article.coverUrl ? (
                  <div className="image">
                    <img
                      src={`${REACT_APP_IMAGE_BASE_URL}/${imageRectangle}/${article.coverUrl}`}
                      alt={article.title}
                    />
                  </div>
                ) : (
                  ''
                )}
              </LazyLoad>

              <div className="container">
                <h1 className="xxlarge-text medium-v-padding">{article.title}</h1>
                <div className="articleInfo">
                  <div className="row">
                    <div className="small-screen-4 medium-screen-2 large-screen-2">
                      <span className="avatar">
                        <img src={avatar} alt={article.title} />
                        <span>{' John Doe'}</span>
                      </span>
                      <Link to="/" className="button small-button primary">
                        Follow
                      </Link>
                      <span className="medium-h-padding">{timeStamp(article.createdAt)}</span>
                    </div>
                  </div>
                </div>
                <br />
                <div className="large-text">{article.description}</div>

                <div className="divider light" />
                <div className="articleBody">
                  {article && <Editor editorState={editorState} readOnly={false} />}
                </div>
              </div>
            </div>
          ) : (
            <div>{!article ? <NotFound /> : <div>{''}</div>}</div>
          )}
        </div>
      </Layout>
    );
  }
}

Article.defaultProps = { match: { params: { slug: '' } } };

Article.propTypes = {
  article: PropTypes.object,
  fetchOneArticle: PropTypes.func.isRequired,
  editorState: PropTypes.func,
  match: PropTypes.object,
  slug: PropTypes.string,
  params: PropTypes.object,
  message: PropTypes.object,
  errors: PropTypes.object
};
const mapStateToProps = ({ articles: { article, errors } }) => ({ article, errors });

export default connect(
  mapStateToProps,
  { fetchOneArticle }
)(Article);
