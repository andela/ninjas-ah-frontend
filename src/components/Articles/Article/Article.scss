@import '../../../assets/css/_padding';
@import '../../../assets/css/_radius';
@import '../../../assets/css/_colors';
@import '../../../assets/css/_shadow';

#article {
  .modal-content {
    background: initial;
    border: 0;
    box-shadow: none;
    .modal-body {
      @extend .white, .radius-2;
      padding: 5px;
    }
  }
  .articleBody {
    line-height: 1.8;
    max-width: 100%;
    ::selection {
      font-weight: bold !important;
      background: lightgreen !important;
    }
    .selection {
      background: lightgreen !important;
      font-style: italic;
      font-weight: bold;
    }

    span[style*='--highlight-id'] {
      background-color: lightgreen;
      cursor: pointer;
      col &:hover {
        opacity: 0.5;
      }
    }
  }
  .articleBody img {
    max-width: 100%;
    margin: 10px auto;
    display: block;
  }
  .articleBody img {
    padding: $small/2;
    background: $light !important;
    @extend .shadow-2;
  }
  .articleBody strong {
    font-weight: 800 !important;
  }
  .articleBody blockquote {
    padding: 20px;
    background: #efefef;
    border-left: 5px solid #343434;
  }
  .articleBody a {
    color: #0b3b94 !important;
  }
  .articleInfo {
    @extend .xxlarge-v-padding;
    .avatar {
      position: relative;
      font-size: ($small * 4/3);
      float: left;
    }
    .avatar img {
      position: absolute;
      top: -8px;
      left: -10px;
      width: 32px;
      height: 32px;
      overflow: hidden;
      border-radius: 50%;
    }
    .avatar span {
      position: relative;
      margin-left: 30px;
      margin-right: 15px;
      top: 0px;
    }
    .small-button {
      padding: $small/3 $small !important;
      @extend .radius-4;
    }
    .rating {
      float: right;
    }
  }

  .tags-values {
    float: left;
    margin: 15px 15px 15px 0;
  }

  .value-tag {
    color: $black;
    background: $lighten-grey;
    padding: 9px;
    border-radius: 16px;
    -webkit-border-radius: 16px;
    -moz-border-radius: 16px;
    -ms-border-radius: 16px;
    -o-border-radius: 16px;
  }
}
