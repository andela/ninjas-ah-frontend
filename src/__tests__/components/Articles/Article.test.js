import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'react-thunk';
import { Provider } from 'react-redux';
import { article, newHighlight } from '../../../__mocks__/article';
import { Article } from '../../../components/Articles/Article/Article';
import { shallow } from '../../../../config/enzymeConfig';

const props = {
  article: {
    id: 1,
    ...article
  },
  match: { params: { slug: 'slug-slug-slug' } },
  fetchOneArticle: jest.fn()
};

describe('<Article />', () => {
  it('should render a <Article /> component', () => {
    const component = shallow(<Article {...props} />);
    component.instance().onChange();
    expect(component).toMatchSnapshot();
  });

  it('should display an article', () => {
    const component = shallow(<Article {...props} />);
    component.instance().displayArticle(props.article);
    component.instance().showHighlights([], JSON.parse(article.body));
    component.instance().showHighlights([newHighlight], JSON.parse(props.article.body));
    expect(component).toMatchSnapshot();
  });
});
