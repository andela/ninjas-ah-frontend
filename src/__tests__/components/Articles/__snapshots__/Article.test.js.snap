// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> should display an article 1`] = `ShallowWrapper {}`;

exports[`<Article /> should render a <Article /> component 1`] = `ShallowWrapper {}`;
